### PR TITLE
Ensure router handles zero length buffers properly

### PR DESCRIPTION
**Pull request by pnarayanan**
Ensure proper handling of zero length buffers in the input
channel within router put operations.

Fixes linkedin/ambry#592

--

Added test cases to handle zero sized buffers at the end and every now and then from the channel.

### CODE REVIEW

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 1]
```diff
@@ -353,8 +353,12 @@ void fillChunks() {
       if (chunkFillingCompletedSuccessfully) {
         PutChunk lastChunk = getBuildingChunk();
         if (lastChunk != null) {
-          lastChunk.onFillComplete(true);
-          updateChunkFillerWaitTimeMetrics();
+          if (chunkCounter != 0 && lastChunk.buf.position() == 0) {
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: `lastChunk.buf.position() == 0` is not a reliable check since buffers can have position > 0 and yet have no data because position == limit.

Also, what happens if there are zero sized buffers in the middle? I think it should be fine - but a test would be useful (if not already there)

Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: As we saw in person, `lastChunk` refers to the `PutChunk` and not the buffer from the `ReadableStreamChannel` (which is named as `channelReadBuffer` in this method). For the `PutChunk` buffer, this check is guaranteed to be correct as this is done only if the chunk is in `Building` state and the method that takes the chunk to that state (`prepareForBuilding()`) sets the position to 0.

For the second part, yes - `testDelayedChannelPutSuccess()` in `PutManagerTest` tests zero-sized last as well as intermediate buffers (probabilistically for the latter).

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java [hunk 2]
```diff
@@ -879,13 +907,16 @@ void verifyNotificationsForFailedPut() {
...
+  private final Random random = new Random();
 
   /**
    * Constructs a MockReadableStreamChannel.
    * @param size the number of bytes that will eventually be written into this channel.
    */
-  MockReadableStreamChannel(long size) {
+  MockReadableStreamChannel(long size, boolean sendZeroSizedLastChunk) {
     this.size = size;
+    this.sendZeroSizedLastChunk = sendZeroSizedLastChunk;
```
Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: This variable might need a more descriptive name since it is doing more than just send zero sized last chunk.

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java [hunk 3]
```diff
@@ -916,16 +947,34 @@ void write(ByteBuffer buf) throws Exception {
...
     readSoFar += toRead;
     if (readSoFar == size) {
+      if (sendZeroSizedLastChunk) {
+        // Send a final zero sized buffer.
+        sendZeroAndWait();
+      }
       returnedFuture.done(size, null);
       callback.onCompletion(size, null);
+    } else {
+      if (sendZeroSizedLastChunk) {
```
Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: minor:
```
if (sendZeroSizedLastChunk && random.nextInt(2) % 2 == 0) {
  sendZeroAndWait();
}  
```

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 4]
```diff
@@ -353,8 +353,12 @@ void fillChunks() {
...
+            logger.trace("The last chunk received from ReadableStreamChannel has no data, discarding");
```
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: last PutChunk in building state having no data is the mark of last chunk or chunks from the `ReadableStreamChannel` having no data. I could rephrase the message to make it more general.

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java (hunk 2 above)
Comment (pnarayanan) [inline ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: fixed.

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java (hunk 3 above)
Comment (pnarayanan) [inline ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: done.

**Comment by pnarayanan** [comment]
@vgkholla Updated. Built, tested and formatted.

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java [hunk 5]
```diff
@@ -879,13 +907,16 @@ void verifyNotificationsForFailedPut() {
   private final long size;
   private long readSoFar;
   private boolean beBad = false;
+  private final boolean sendZeroSizedBuffers;
+  private final Random random = new Random();
 
   /**
    * Constructs a MockReadableStreamChannel.
    * @param size the number of bytes that will eventually be written into this channel.
```
Comment (cgtz) [inline ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: add a javadoc for `sendZeroSizedBuffers`

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java [hunk 6]
```diff
@@ -379,24 +380,39 @@ public void testLaterChunkFailure() throws Exception {
...
+    for (int blobSize : blobSizes) {
+      testDelayed(blobSize, false);
+      testDelayed(blobSize, true);
+    }
+    assertSuccess();
+    assertCloseCleanup();
+  }
+
+  /**
+   * Helper method to put a blob with a channel that does not have all the data at once.
```
Comment (cgtz) [inline ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: Minor: could you add parameter docs for this method

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java [hunk 7]
```diff
@@ -379,24 +380,39 @@ public void testLaterChunkFailure() throws Exception {
   }
 
   /**
-   * Tests put of a blob with a channel that does not have all the data at once.
+   * Tests put of blobs with various sizes with channels that do not have all the data at once.
```
Comment (cgtz) [inline ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: Maybe document that this also tests sending empty chunks.

**Comment by pnarayanan** [comment]
@cgtz @vgkholla Addressed comments.
